plugins/funind/g_indfun.mlg

```diff
@@ -173,24 +173,32 @@ let () =
   let raw_printer env sigma _ _ _ (loc,body) = Ppvernac.pr_rec_definition body in
   Pptactic.declare_extra_vernac_genarg_pprule wit_function_rec_definition_loc raw_printer
 
+let hard_function recsl =
+  List.exists (function
+      | _,((_,(Some { CAst.v = CMeasureRec _ }
+              | Some { CAst.v = CWfRec _}),_,_,_),_) -> true
+      | _,((_,Some { CAst.v = CStructRec _ },_,_,_),_)
+      | _,((_,None,_,_,_),_) -> false) recsl
+
+let classify_function recsl =
+  let hard = hard_function recsl in
+  match
+    Vernac_classifier.classify_vernac
+      (Vernacexpr.(CAst.make @@ VernacExpr([], VernacFixpoint(Decl_kinds.NoDischarge, List.map snd recsl))))
+  with
+  | Vernacextend.VtSideff ids, _ when hard ->
+    Vernacextend.(VtStartProof (GuaranteesOpacity, ids), VtLater)
+  | x -> x
 }
 
 (* TASSI: n'importe quoi ! *)
-VERNAC COMMAND EXTEND Function
-| ![ proof ] ["Function" ne_function_rec_definition_loc_list_sep(recsl,"with")]
-    => { let hard = List.exists (function
-           | _,((_,(Some { CAst.v = CMeasureRec _ }
-             | Some { CAst.v = CWfRec _}),_,_,_),_) -> true
-           | _,((_,Some { CAst.v = CStructRec _ },_,_,_),_)
-           | _,((_,None,_,_,_),_) -> false) recsl in
-         match
-           Vernac_classifier.classify_vernac
-             (Vernacexpr.(CAst.make @@ VernacExpr([], VernacFixpoint(Decl_kinds.NoDischarge, List.map snd recsl))))
-         with
-         | Vernacextend.VtSideff ids, _ when hard ->
-             Vernacextend.(VtStartProof (GuaranteesOpacity, ids), VtLater)
-         | x -> x }
```

Comment:
Apparently this classification is bullshit, in bug_1618.v the second Function is classified as startproof but doesn't start a proof.
What should we do?

Reply:
IMHO at this level the API doesn't have a problem with a conditional proof open, I would just reintroduce `VtMaybeOpenProof` and remove the static hack.

Reply:
I do dislike the `Option.get`, the whole point of this refactoring is to avoid such stuff.

Reply:
Moreover, clients of the vernacinterp API should not have to rely on classification, which is an implementation detail of the STM.

Reply:
I would much prefer if we kept parsing and execution separated. As of today, opening a proof has an impact on parsing. It was a lot of work to remove `VtUnknown`, please don't reintroduce a `VtMaybeOpenProof`. Let's try to figure out what `Function` does instead.

Reply:
I am not sure I follow what you mean, parsing and execution cannot be separated. For example, what about `Notation` ?

`VtUnknown` is specific to the STM, and it bears little relation to this layer IMHO. I am more of the opinion that the STM should adapt to the demands of the system instead of the system having to adapt to its current limitations.

Reply:
> I am not sure I follow what you mean, parsing and execution cannot be separated. For example, what about `Notation` ?

We had this discussion many times before, including in the presentation I gave at the last WG.

> `VtUnknown` is specific to the STM, and it bears little relation to this layer IMHO. I am more of the opinion that the STM should adapt to the demands of the system instead of the system having to adapt to its current limitations.

I guess it depends if the behavior of the system is reasonable (for some notion of reasonable). Let's start with describing it, maybe: when does `Function` open a proof?

Reply:
This has nothing to do with "keeping parsing and execution separate", but with static invariants. The interp layer has not problem dealing with a conditional proof open, then the STM can use its static analysis to infer what it does. However this piece of code [in the current PR] is not acceptable:
```
let static_maybe_open_proof f = function
  | true -> VtOpenProof (fun () -> Option.get (f ()))
  | false -> VtDefault (fun () -> let o = f () in assert (Option.is_empty o))
```
we kinda defeat the purpose of the work, which is to ensure properties statically, however here we have two potential anomalies.

Reply:
> This has nothing to do with "keeping parsing and execution separate"

I disagree, for the reasons I mentioned above.

> However this piece of code [in the current PR] is not acceptable:

Ok, let's try to improve it. When does `Function` open a proof?

Reply:
> I disagree, for the reasons I mentioned above.

What reasons? I am not sure I follow. What you do for `Require` for example?

Reply:
> What do you do for `Require` for example?

We discussed this before too. In fact, it was the reason I started the libobject refactoring. If your memory needs to be refreshed, I'll write down the plan somewhere, but let's not pollute this PR too much.

Reply:
>I would much prefer if we kept parsing and execution separated. As of today, opening a proof has an impact on parsing. It was a lot of work to remove VtUnknown, please don't reintroduce a VtMaybeOpenProof. Let's try to figure out what Function does instead.

Let's keep in mind thay maybeopenproof is more precise than the current `stack option -> stack option`